src/main/java/graphql/util/TraverserContext.java

```diff
@@ -0,0 +1,80 @@
+package graphql.util;
+
+import graphql.Internal;
+
+import java.util.Set;
+
+/**
+ * Traversal context
+ *
+ * @param <T> type of tree node
+ */
+@Internal
+public interface TraverserContext<T> {
+    /**
+     * Returns current node being visited
+     *
+     * @return current node traverser is visiting
+     */
+    T thisNode();
+
+    /**
+     * Returns parent context.
+     * Effectively organizes Context objects in a linked list so
+     * by following {@link #parentContext() } links one could obtain
+     * the current path as well as the variables {@link #getVar(java.lang.Class) }
+     * stored in every parent context.
+     *
+     * Useful when it is difficult to organize a local Visitor's stack, when performing
+     * breadth-first or parallel traversal
+     *
+     * @return context associated with the node parent
+     */
+    TraverserContext<T> parentContext();
+
+    Object getParentResult();
+
+    /**
+     * Informs that the current node has been already "visited"
+     *
+     * @return {@code true} if a node had been already visited
+     */
+    boolean isVisited();
+
+    /**
+     * Obtains all visited nodes and values received by the {@link TraverserVisitor#enter(graphql.util.TraverserContext, java.lang.Object) }
+     * method
+     *
+     * @return a map containg all nodes visited and values passed when visiting nodes for the first time
+     */
+    Set<T> visitedNodes();
+
+    /**
+     * Obtains a context local variable
+     *
+     * @param <S> type of the variable
+     * @param key key to lookup the variable value
+     *
+     * @return a variable value of {@code null}
+     */
+    <S> S getVar(Class<? super S> key);
+
+    /**
+     * Stores a variable in the context
+     *
+     * @param <S>   type of a varable
+     * @param key   key to create bindings for the variable
+     * @param value value of variable
+     *
+     * @return this context to allow operations chaining
+     */
+    <S> TraverserContext<T> setVar(Class<? super S> key, S value);
+
+
+    void setResult(Object result);
+
+    Object getResult();
```

Comment:
Could you share your thoughts about making getResult/setResult/getInitialData of an object type?

Reply:
First reason was to make `TraverserVisitor` and `NodeVisitor` more clear about the return type: Before that it was just `Object` which meant anything or a special Marker Enum. Now it is clear that you should return a enum and only a enum which controls the general Visitor (Abort, Exit etc). 

Second reason was that passing some data along and maybe generating new data out of it is a special case of the general Context imho: the last result generated (the parent result) can be used to generate a new result. But not always and I didn't found a good reason to introduce a extra argument just for this case. Now `TraverserContext` encapsulates really the whole context.

Reply:
I had a similar comment on the NodeTraverser.

When a Visitor is used by a `Traverser`, it is most likely the argument to the `Visitor` will be `TraverserContext`. Therefore in my opinion it would be cleaner and, actually, more efficient to return from the Visitor a TraverserContext. Visitor can set on the Traverser a result as well a `TraverserControl` value. If we do this, Visitor's definition would look like:
```java
interface NodeVisitor<T> {
   T visitField (Field node, T data);
   ...
}
```
And this will allow to simplify simple double dispatch case, where I there is no need to traverse a tree, ,but perform an action based on the node runtime type.